stock_helper/models/stock_location.py

```diff
@@ -18,3 +19,25 @@ def is_sublocation_of(self, others, func=any):
         # Efficient way to verify that the current location is
         # below one of the other location without using SQL.
         return func(self.parent_path.startswith(other.parent_path) for other in others)
+
+    def get_closest_warehouse(self):
+        """Returns warehouse id of the closest warehouse that contains location
+
+        By default the get_warehouse (which is located in the odoo core module stock)
+        returns the warehouse via searching the view_location_id with parent_of.
+        If we have multiple warehouses
+        where a view_location_id is a parent of an other view_location_id
+        Then it depends on the sorting of the warehouses which this method returns
+
+        With this methods we will really get the closest warehouse of a location
+        """
+        self.ensure_one()
+        warehouses = (
+            self.env["stock.warehouse"]
+            .search([])
+            .sorted(lambda w: w.view_location_id.parent_path, reverse=True)
+        )
```

Comment:
can we optimized this by filtering only on relevant WHs?
```suggestion
        location_ids = [int(x) for x in self.parent_path.split("/")]
        warehouses = (
            self.env["stock.warehouse"]
            .search([("view_location_id", "in", location_ids)])
            .sorted(lambda w: w.view_location_id.parent_path, reverse=True)
        )
```